<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [3]:
import pandas as pd
import psycopg2

In [4]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

DBNAME = 'project_sql'
USER = '**********'
PASSWORD = '************'
HOST = '*************'
PORT = int("*****")

In [5]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [6]:
# текст запроса
query_3_1 = f'''
                select count(*) as sum_vacancies
                from vacancies v
            '''

In [7]:
# результат запроса
sum_vacancies = pd.read_sql_query(query_3_1, connection)
sum_vacancies

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\3348228061.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  sum_vacancies = pd.read_sql_query(query_3_1, connection)


sum_vacancies
0          49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [8]:
# текст запроса
query_3_2 = f'''select count(*) as sum_employers
                from employers
            '''

In [9]:
# результат запроса
sum_employers = pd.read_sql_query(query_3_2, connection)
sum_employers

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\2255150949.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  sum_employers = pd.read_sql_query(query_3_2, connection)


sum_employers
0          23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [10]:
# текст запроса
query_3_3 = f'''select count(id) as quantity_areas
                from areas a
            '''

In [11]:
# результат запроса
quantity_areas = pd.read_sql_query(query_3_3, connection)
quantity_areas

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\2486803467.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  quantity_areas = pd.read_sql_query(query_3_3, connection)


quantity_areas
0            1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [12]:
# текст запроса
query_3_4 = f'''select count(id) as industry_counter
                from industries i
            '''

In [13]:
# результат запроса
industry_counter = pd.read_sql_query(query_3_4, connection)
industry_counter

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\2225315761.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  industry_counter = pd.read_sql_query(query_3_4, connection)


industry_counter
0               294

***

### выводы по предварительному анализу данных

В базе данных содержится следующая информация:
    
    - 49197 вакансий 
    
    - 23501 работодателей
    
    - расположены вакансии в 1362 регионах
    
    - вакансии представленны в 294 различных сферах деятельности 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [14]:
# текст запроса
query_4_1 = '''
    select a.name as area, count(v.id) as cnt
    from vacancies v 
        join areas a on v.area_id = a.id
    group by a.name
    order by cnt desc, area
    limit 10
    '''

In [15]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892
5     Екатеринбург  1698
6  Нижний Новгород  1670
7           Казань  1415
8        Краснодар  1301
9           Самара  1144

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [16]:
# текст запроса
query_4_2 = '''
    select count(v.id) as cnt
    from vacancies v 
    where salary_from is not null or salary_to is not null
    '''

In [17]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [18]:
# текст запроса
query_4_3 = '''
SELECT 
AVG(salary_from) salary_down,
AVG(salary_to) salary_up
  FROM vacancies v
'''

In [19]:
# результат запроса
query_4_3_df = pd.read_sql_query(query_4_3, connection)
print(round(query_4_3_df['salary_down'][0], 0))
print(round(query_4_3_df['salary_up'][0], 0))

71065.0
110537.0


C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\2514329214.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_4_3_df = pd.read_sql_query(query_4_3, connection)


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [20]:
# текст запроса
query_4_4 = '''select count(id) as vacancies_count, schedule, employment
            from vacancies
            group by schedule, employment
            order by vacancies_count desc
            offset 1
            limit 1
'''

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


vacancies_count          schedule        employment
0             7802  Удаленная работа  Полная занятость

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [22]:
# текст запроса
query_4_5 = '''select experience, count(id) as vacancies_count
            from vacancies
            group by experience
            order by vacancies_count 
'''

In [23]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  vacancies_count
0         Более 6 лет             1337
1           Нет опыта             7197
2       От 3 до 6 лет            14511
3  От 1 года до 3 лет            26152

***

### выводы по детальному анализу вакансий

    - Количество вакансий представлено для опыта работы от 1 года до 3 лет: 26152 шт.

    - Количество вакансий - для специалистов с опытом работы более 6 лет: 1337 шт.

    - Наибольшее количество вакансий в городах-милиоонниках.

    - Поиск работы возможен по типу рабочего графика и типу трудоустройства.

    - Присутствуют различные вакансии: без опыта работы, с опытом работы от 1 до 3 лет, от 3 до 6 лет и с опытом работы более 6 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [24]:
# текст запроса
query_5_1 = '''select e.name as employer, count(v.id) as vacancies_count
            from vacancies v join employers e on v.employer_id = e.id
            group by employer
            order by vacancies_count desc
            limit 5
'''

In [25]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


employer  vacancies_count
0         Яндекс             1933
1     Ростелеком              491
2       Тинькофф              444
3           СБЕР              428
4  Газпром нефть              331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [26]:
# текст запроса
query_5_2 = '''select a.name as region_name,
                       count(e.id) as total_employers,
                       count(v.id) as total_vacancies
                from areas a
                left join employers e
                on a.id = e.area
                left join vacancies v
                on a.id = v.area_id
                where v.id is null
                group by region_name
                order by total_employers DESC
                limit 1
            '''

In [27]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


region_name  total_employers  total_vacancies
0      Россия              410                0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [28]:
# текст запроса
query_5_3 = '''select e.name,
                       count(distinct a.id) as no_of_regions
                from vacancies v
                join areas a
                on v.area_id = a.id
                join employers e
                on v.employer_id = e.id
                group by e.name
                order by no_of_regions DESC
                limit 5
            '''

In [29]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  no_of_regions
0                  Яндекс            181
1              Ростелеком            152
2              Спецремонт            116
3  Поляков Денис Иванович             88
4               ООО ЕФИН              71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [30]:
# текст запроса
query_5_4 = '''select count(e.name) as employers_count
            from employers e
            left join employers_industries i on e.id = i.employer_id
            where i.industry_id is null
'''

In [31]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
display(df)

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\375529716.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


employers_count
0             8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [32]:
# текст запроса
query_5_5 = '''select e.name  
            from employers e
            left join employers_industries i on e.id = i.employer_id
            group by e.name
            having count(i.industry_id) = 4
            order by 1
            offset 2
            limit 1
'''

In [33]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
# текст запроса
query_5_6 = '''select count(e.id)
            from employers e
            join employers_industries ei on e.id = ei.employer_id
            join industries i on i.id = ei.industry_id
            where i.name = 'Разработка программного обеспечения'
'''

In [35]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
display(df)

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\1761488907.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [37]:
# код для получения списка городов-милионников

import requests
from bs4 import BeautifulSoup

url = "https://ru.wikipedia.org/wiki/Список_городов_России_с_населением_более_100_тысяч_жителей"
response = requests.get(url)

if response.status_code == 200:
    soup = BeautifulSoup(response.text, "html.parser")
    cities = []

    for row in soup.select("table.wikitable tr")[1:]:
        cells = row.find_all("td")
        if len(cells) > 2:
            city = cells[2].text.strip()
            try:
                population = (
                    int(cells[-1].text.strip().replace("\xa0", "").replace(" ", ""))
                    * 1000
                )
                if population >= 1000000:
                    cities.append(city)
            except ValueError:
                continue

    print("\n".join(cities))



Москва
Санкт-Петербург
Новосибирск
Екатеринбург
Казань
Красноярск
Нижний Новгород
Челябинск
Уфа
Самара
Ростов-на-Дону
Краснодар
Омск
Воронеж
Пермь
Волгоград


In [44]:
# текст запроса
# Преобразуем список городов в формат, подходящий для SQL
cities_tuple = tuple(cities)
cities_str = str(cities_tuple)

# Если в списке только один город, добавим запятую после города, чтобы SQL распознал это как кортеж
if len(cities) == 1:
    cities_str = f"('{cities[0]}',)"

query_5_7 = f''' 
select 
a.name,
count(v.id) cnt
  from vacancies v 
  join employers e ON e.id=v.employer_id
  join areas a ON a.id=v.area_id
  where e.name = 'Яндекс' AND a.name IN {cities_str}
group by a.name  

union all
select
'Total', count(v.id)
  from vacancies v 
  join employers e ON e.id=v.employer_id
  join areas a ON a.id=v.area_id
  where e.name = 'Яндекс' AND a.name IN {cities_str}
group by e.name
order by cnt 
'''


In [45]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

In [ ]:
# выводы по анализу работодателей
Компания Яндекс является лидером по количеству вакансий. 
Яндекс на первом месте по охвату регионов.
Существуют регионы, в которых нет вакансий, несмотря на большое количество работодателей в них. 
В БД существует 8419 работодателей, у которых не указана сфера деятельности. 
Существует 3553 работодателя, у которых в качестве сферы деятельности указана "Разработка ПО".


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [46]:
# текст запроса
query_6_1 = '''select count(id) as vacancies_count
           from vacancies  
           where lower(name) like '%data%' or lower(name) like '%данн%' 
'''

In [47]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


vacancies_count
0             1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [63]:
# текст запроса
query_6_2 = '''select count(id) as vacancies_count
           from vacancies  
           where (lower(name) like '%data scientist%' 
            or lower(name) like '%data science%' 
            or lower(name) like '%исследователь данных%'
            or (name like '%ML%' and name not like '%HTML%')
            or lower(name) like '%machine learning%'
            or lower(name) like '%машинн%обучен%')
            and (lower(name) like '%junior%'
                or lower(experience) like '%нет опыта%' 
                or lower(employment) like '%стажировка%'
                )
'''

In [64]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


vacancies_count
0               51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [66]:
# текст запроса
query_6_3 = '''select count(v.id) as sql_preferred
                from vacancies v
                
                where (lower(v.name) like '%data scientist%'
                or lower(v.name) like '%data science%'
                or lower(v.name) like '%исследователь данных%'
                
                or (v.name like '%ML%'and v.name not like '%HTML%')
                
                or lower(v.name) like '%machine learning%'
                or lower(v.name) like '%машинн%обучен%')
                
                and (lower(v.key_skills) like '%sql%'
                or lower(v.key_skills) like '%postgres%')
            '''

In [67]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


sql_preferred
0            201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [68]:
# текст запроса
query_6_4 = '''select count(id) as vacancies_count
           from vacancies  
           where (lower(name) like '%data scientist%' 
            or lower(name) like '%data science%' 
            or lower(name) like '%исследователь данных%'
            or (name like '%ML%' and name not like '%HTML%')
            or lower(name) like '%machine learning%'
            or lower(name) like '%машинн%обучен%')
            and lower(key_skills) like '%python%'
'''

In [69]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


vacancies_count
0              351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [70]:
# текст запроса
query_6_5 = '''select round(avg(array_length(regexp_split_to_array(key_skills, '\t'), 1)), 2) as mean_skills_count
           from vacancies  
           where (lower(name) like '%data scientist%' 
            or lower(name) like '%data science%' 
            or lower(name) like '%исследователь данных%'
            or (name like '%ML%' and name not like '%HTML%')
            or lower(name) like '%machine learning%'
            or lower(name) like '%машинн%обучен%')
'''

In [71]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


mean_skills_count
0               6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [72]:
# текст запроса
query_6_6 = '''
    select experience, 
        round(avg((coalesce(salary_from, 0) + coalesce(salary_to, 0)) / case when salary_from > 0 AND salary_to > 0 THEN 2 ELSE 1 END)) as mean_salary 
    from vacancies  
    where (lower(name) like '%data scientist%' 
            or lower(name) like '%data science%' 
            or lower(name) like '%исследователь данных%'
            or (name like '%ML%' and name not like '%HTML%')
            or lower(name) like '%machine learning%'
            or lower(name) like '%машинн%обучен%')
            and (salary_from IS NOT NULL OR salary_to IS NOT NULL)
    group by experience
'''

In [73]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  mean_salary
0           Нет опыта      74643.0
1  От 1 года до 3 лет     139675.0
2       От 3 до 6 лет     243115.0

***

In [ ]:
# выводы по предметному анализу
В базе представлена 1771 вакансия, имеющая отношение к данным. 
Из них только 51 вакансия подойдет для начинающих дата-сайентистов. 
В 201 DS-вакансии в качестве ключевого навыка указан SQL и Postgres. 
В 351 вакансии Python указан в качестве ключевого навыка DS. 
Среди вакансий по DS в среднем от соискателя требуется 6.41 ключевых навыков по специальности. 
Согласно представленным по DS вакансиям, средняя зарплата DS-специалиста стремительно растет с опытом работы.


# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

### Дополнительный Анализ

### Средняя зарплата по городам-миллионикам

In [75]:
query_7_1 = f'''select a.name as region_name,
                       ceil(avg(v.salary_from)) as salary_start,           
                       ceil(avg(v.salary_to)) as salary_end,
                       ceil(avg(((coalesce(v.salary_from, v.salary_to) + coalesce(v.salary_to, v.salary_from)) / 2))) as average_salary
                from vacancies v
                
                left join areas a
                on v.area_id = a.id
                
                where a.name in {cities_str}
              
                group by region_name
                
                order by average_salary DESC
            '''

In [76]:
result_1 = pd.read_sql_query(query_7_1, connection)
result_1

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\2879213420.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_1 = pd.read_sql_query(query_7_1, connection)


region_name  salary_start  salary_end  average_salary
0            Москва      113410.0    160595.0        134141.0
1       Новосибирск       91272.0    137573.0        107777.0
2   Санкт-Петербург       89893.0    127249.0        104410.0
3   Нижний Новгород       83734.0    126921.0         96059.0
4             Пермь       78933.0    130215.0         95985.0
5    Ростов-на-Дону       79392.0    119709.0         94066.0
6         Краснодар       78484.0    117144.0         92329.0
7            Казань       76118.0    119942.0         92208.0
8      Екатеринбург       77489.0    120404.0         91896.0
9              Омск       74899.0    107686.0         86393.0
10          Воронеж       69203.0    108998.0         84959.0
11           Самара       67885.0    101157.0         80224.0
12        Челябинск       69564.0     93331.0         77463.0
13       Красноярск       64130.0     90962.0         75156.0
14              Уфа       62639.0    103066.0         74734.0
15        Волгоград       61014.0     93073.0         68689.0

 ### Tоп-10 средней заработной платы по регионам

In [77]:
query_7_2 = '''select a.name as region,
                       ceil(avg(((coalesce(v.salary_from, v.salary_to) + coalesce(v.salary_to, v.salary_from)) / 2))) as average_salary
                from vacancies v
                
            left join areas a
            on v.area_id = a.id
              
            group by region
            
            --filter by non-null salary amounts only
            having ceil(avg(((coalesce(v.salary_from, v.salary_to) + coalesce(v.salary_to, v.salary_from)) / 2))) > 0
            
            order by 2 DESC
            
            --return top 10
            limit 10
            '''

In [79]:
result_2 = pd.read_sql_query(query_7_2, connection)
result_2

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\2922074590.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_2 = pd.read_sql_query(query_7_2, connection)


region  average_salary
0                            Германия        452751.0
1                            Болгария        324348.0
2                             Испания        300000.0
3                          Черногория        288590.0
4                               Литва        263019.0
5                              Польша        257156.0
6  Петропавловка (Республика Бурятия)        250000.0
7                              Турция        235436.0
8                                Кипр        223736.0
9                             Кутаиси        220475.0

### Топ-10 отраслей по средней зарплате

In [80]:
query_7_3 = f'''select i.name as profession,
                       ceil(avg(((coalesce(v.salary_from, v.salary_to) + coalesce(v.salary_to, v.salary_from)) / 2))) as average_salary
                from vacancies v
                
                left join employers e
                on v.employer_id = e.id
                
                left join employers_industries ei
                on e.id = ei.employer_id
                
                left join industries i
                on ei.industry_id = i.id
                
                where ei.industry_id is not null
                
                and (v.salary_from is not null
                or v.salary_to is not null)
                
                group by profession
                
                order by 2 DESC
                
                limit 10
            '''

In [82]:
result_3 = pd.read_sql_query(query_7_3, connection)
result_3

C:\Users\Chaos\AppData\Local\Temp\ipykernel_5872\459209836.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result_3 = pd.read_sql_query(query_7_3, connection)


profession  average_salary
0          Агрохимия (продвижение, оптовая торговля)        381429.0
1                           Агрохимия (производство)        339875.0
2                                     Игорный бизнес        191905.0
3                  Корма для животных (производство)        168300.0
4                                         Зоомагазин        167613.0
5                                 Кадровые агентства        144530.0
6      Лакокрасочная продукция, сырье (производство)        142803.0
7  Бытовая химия, парфюмерия, косметика  (произво...        130120.0
8  Лакокрасочная продукция, сырье (продвижение, о...        128445.0
9  Дорожно-строительная техника, сельскохозяйстве...        122878.0